File: py-basemapper/src/basemapper/providers.py
"""Provider Generator Helpers — convert bare tile URLs to MapLibre GL Style JSON."""

from __future__ import annotations

import json
import warnings
from typing import Dict, Optional

_FILL_KEYS = {"fill-color", "fill-opacity", "fill-outline-color"}
_LINE_KEYS = {"line-color", "line-width", "line-opacity"}
_KNOWN_KEYS = _FILL_KEYS | _LINE_KEYS


class RasterProvider:
    """Generate a MapLibre GL Style JSON string for an XYZ raster tile source.

    Cast to ``str`` to obtain the inline MapLibre GL Style JSON.

    Args:
        url_template: Slippy-map tile URL with ``{z}``, ``{x}``, ``{y}`` placeholders.
        tile_size: Tile size in pixels (default 256).
    """

    def __init__(self, url_template: str, tile_size: int = 256) -> None:
        self.url_template = url_template
        self.tile_size = tile_size

    def to_style_json(self) -> str:
        """Return a MapLibre GL Style JSON string."""
        style = {
            "version": 8,
            "sources": {
                "raster-source": {
                    "type": "raster",
                    "tiles": [self.url_template],
                    "tileSize": self.tile_size,
                }
            },
            "layers": [
                {
                    "id": "raster-layer",
                    "type": "raster",
                    "source": "raster-source",
                }
            ],
        }
        return json.dumps(style)

    def __str__(self) -> str:
        return self.to_style_json()


class VectorProvider:
    """Generate a MapLibre GL Style JSON string for an MVT vector tile source.

    Accepted paint keys: ``fill-color``, ``fill-opacity``, ``fill-outline-color``,
    ``line-color``, ``line-width``, ``line-opacity``. Unrecognised keys are dropped
    with a warning. Omitting ``paint`` applies a light-grey default style.

    Args:
        url_template: MVT tile URL with ``{z}``, ``{x}``, ``{y}`` placeholders.
        paint: Optional dict of MapLibre GL paint properties.
    """

    def __init__(self, url_template: str, paint: Optional[Dict[str, object]] = None) -> None:
        self.url_template = url_template
        self.paint = paint or {}

    def to_style_json(self) -> str:
        paint = dict(self.paint)
        unknown = set(paint.keys()) - _KNOWN_KEYS
        for key in unknown:
            warnings.warn(
                f"VectorProvider: unrecognised paint key '{key}' dropped.",
                stacklevel=2,
            )
            del paint[key]

        fill_paint: Dict[str, object] = {
            k: paint[k] for k in _FILL_KEYS if k in paint
        }
        line_paint: Dict[str, object] = {
            k: paint[k] for k in _LINE_KEYS if k in paint
        }

        if "fill-color" not in fill_paint:
            fill_paint["fill-color"] = "#e8e0d8"
        if "fill-opacity" not in fill_paint:
            fill_paint["fill-opacity"] = 1
        if "line-color" not in line_paint:
            line_paint["line-color"] = "#aaaaaa"
        if "line-width" not in line_paint:
            line_paint["line-width"] = 1

        style = {
            "version": 8,
            "sources": {
                "vector-source": {
                    "type": "vector",
                    "tiles": [self.url_template],
                }
            },
            "layers": [
                {
                    "id": "vector-fill",
                    "type": "fill",
                    "source": "vector-source",
                    "paint": fill_paint,
                },
                {
                    "id": "vector-line",
                    "type": "line",
                    "source": "vector-source",
                    "paint": line_paint,
                },
            ],
        }
        return json.dumps(style)

    def __str__(self) -> str:
        return self.to_style_json()


class EsriVectorProvider:
    """Return the root.json style endpoint URL for an ArcGIS VectorTileServer.

    The Rust core fetches this URL to obtain the full, ESRI-published MapLibre
    GL style JSON. Cast to ``str`` to get the URL string (not inline JSON).

    Args:
        base_url: Base ArcGIS VectorTileServer URL (trailing slashes stripped).
    """

    def __init__(self, base_url: str) -> None:
        self.base_url = base_url.rstrip("/")

    def __str__(self) -> str:
        return f"{self.base_url}/resources/styles/root.json"


class EsriRasterProvider:
    """Generate a MapLibre GL Style JSON string for an ArcGIS MapServer raster source.

    Args:
        url_template: ArcGIS MapServer tile URL (e.g., ``.../MapServer/tile/{z}/{y}/{x}``).
        tile_size: Tile size in pixels (default 256).
    """

    def __init__(self, url_template: str, tile_size: int = 256) -> None:
        self.url_template = url_template
        self.tile_size = tile_size

    def to_style_json(self) -> str:
        style = {
            "version": 8,
            "sources": {
                "esri-raster-source": {
                    "type": "raster",
                    "tiles": [self.url_template],
                    "tileSize": self.tile_size,
                }
            },
            "layers": [
                {
                    "id": "esri-raster-layer",
                    "type": "raster",
                    "source": "esri-raster-source",
                }
            ],
        }
        return json.dumps(style)

    def __str__(self) -> str:
        return self.to_style_json()

File: r-basemapper/R/providers.R
#' Generate a MapLibre GL Style JSON string for an XYZ raster tile source.
#'
#' @param url_template Character: tile URL with `{z}`, `{x}`, `{y}` placeholders.
#' @param tile_size Integer tile size in pixels (default 256L).
#' @return A JSON character string ready to pass to `render_basemap_raw()`.
#' @export
#' @examples
#' style <- raster_provider("https://tile.openstreetmap.org/{z}/{x}/{y}.png")
raster_provider <- function(url_template, tile_size = 256L) {
  style <- list(
    version = 8L,
    sources = list(
      `raster-source` = list(
        type     = "raster",
        tiles    = list(url_template),
        tileSize = as.integer(tile_size)
      )
    ),
    layers = list(
      list(id = "raster-layer", type = "raster", source = "raster-source")
    )
  )
  jsonlite::toJSON(style, auto_unbox = TRUE)
}

#' Return the root.json style endpoint URL for an ArcGIS VectorTileServer.
#'
#' The Rust core fetches this URL to obtain the full, ESRI-published MapLibre
#' GL style JSON. Returns a plain URL string, not inline JSON.
#'
#' @param base_url Character: base ArcGIS VectorTileServer URL.
#' @return A URL character string.
#' @export
#' @examples
#' url <- esri_vector_provider("https://basemaps.arcgis.com/arcgis/rest/services/World_Basemap_v2/VectorTileServer")
esri_vector_provider <- function(base_url) {
  paste0(sub("/*$", "", base_url), "/resources/styles/root.json")
}

#' Generate a MapLibre GL Style JSON string for an ArcGIS MapServer raster source.
#'
#' @param url_template Character: ArcGIS tile URL (e.g., `.../MapServer/tile/{z}/{y}/{x}`).
#' @param tile_size Integer tile size in pixels (default 256L).
#' @return A JSON character string ready to pass to `render_basemap_raw()`.
#' @export
#' @examples
#' style <- esri_raster_provider("https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}")
esri_raster_provider <- function(url_template, tile_size = 256L) {
  style <- list(
    version = 8L,
    sources = list(
      `esri-raster-source` = list(
        type     = "raster",
        tiles    = list(url_template),
        tileSize = as.integer(tile_size)
      )
    ),
    layers = list(
      list(id = "esri-raster-layer", type = "raster", source = "esri-raster-source")
    )
  )
  jsonlite::toJSON(style, auto_unbox = TRUE)
}

#' Generate a MapLibre GL Style JSON string for an MVT vector tile source.
#'
#' Recognised paint keys: `fill-color`, `fill-opacity`, `fill-outline-color`,
#' `line-color`, `line-width`, `line-opacity`. Unrecognised keys are dropped
#' with a `warning()`. An empty paint list applies a light-grey default style.
#'
#' @param url_template Character: MVT tile URL with `{z}`, `{x}`, `{y}` placeholders.
#' @param source_layer Character: the name of the layer within each MVT tile to
#'   render. This is tile-server specific (e.g. `"water"`, `"roads"`); it maps
#'   to the MapLibre GL `"source-layer"` property and is required for vector
#'   sources.
#' @param paint Named list of MapLibre GL paint properties.
#' @return A JSON character string ready to pass to `render_basemap_raw()`.
#' @export
#' @examples
#' style <- vector_provider(
#'   "https://example.com/tiles/{z}/{x}/{y}.mvt",
#'   source_layer = "land",
#'   paint = list("fill-color" = "#e8e0d8", "line-color" = "#aaa")
#' )
vector_provider <- function(url_template, source_layer, paint = list()) {
  fill_keys <- c("fill-color", "fill-opacity", "fill-outline-color")
  line_keys <- c("line-color", "line-width", "line-opacity")
  known_keys <- c(fill_keys, line_keys)

  unknown <- setdiff(names(paint), known_keys)
  if (length(unknown) > 0) {
    warning(paste0("vector_provider: unrecognised paint key(s) dropped: ",
                   paste(unknown, collapse = ", ")))
    paint[unknown] <- NULL
  }

  fill_paint <- paint[intersect(names(paint), fill_keys)]
  line_paint <- paint[intersect(names(paint), line_keys)]

  if (is.null(fill_paint[["fill-color"]]))   fill_paint[["fill-color"]]   <- "#e8e0d8"
  if (is.null(fill_paint[["fill-opacity"]])) fill_paint[["fill-opacity"]] <- 1
  if (is.null(line_paint[["line-color"]]))   line_paint[["line-color"]]   <- "#aaaaaa"
  if (is.null(line_paint[["line-width"]]))   line_paint[["line-width"]]   <- 1

  style <- list(
    version = 8L,
    sources = list(
      `vector-source` = list(
        type  = "vector",
        tiles = list(url_template)
      )
    ),
    layers = list(
      list(id = "vector-fill", type = "fill",
           source = "vector-source", `source-layer` = source_layer,
           paint = fill_paint),
      list(id = "vector-line", type = "line",
           source = "vector-source", `source-layer` = source_layer,
           paint = line_paint)
    )
  )
  jsonlite::toJSON(style, auto_unbox = TRUE)
}
